pkg/cli/ignition/convert3.go

```diff
@@ -0,0 +1,116 @@
+package ignition
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/coreos/ign-converter/translate/v24tov31"
+	ignv2 "github.com/coreos/ignition/config/v2_4"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	kcmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/util/templates"
+)
+
+const convert3CommandName = "convert3"
+
+type convert3Options struct {
+	// Source file
+	Source string
+	// Destination file
+	Destination string
+
+	SourceReader io.Reader
+	DestWriter   io.Writer
+}
+
+var convert3Example = templates.Examples(`
+	# Convert an Ignition Spec 2 pointer configuration to Spec 3
+	oc ignition convert3 < worker.ign > worker.ign3
+`)
+
+// NewCommandConvert3 creates the command
+func NewCommandConvert3(streams genericclioptions.IOStreams) *cobra.Command {
+	o := convert3Options{
+		SourceReader: streams.In,
+		DestWriter:   streams.Out,
+	}
+	cmd := &cobra.Command{
+		Use:     convert3CommandName,
+		Short:   "Convert Ignition Spec 2 to Spec 3",
+		Example: convert3Example,
+		Args:    cobra.ExactArgs(0),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(o.validate(args))
+			kcmdutil.CheckErr(o.run())
+		},
+	}
+
+	cmd.Flags().StringVar(&o.Source, "in", o.Source, "File containing input Ignition to convert. Read from stdin if omitted.")
+	cmd.Flags().StringVar(&o.Destination, "out", o.Destination, "Path to file for converted Ignition. Written to stdout if omitted.")
+	// autocompletion hints
+	cmd.MarkFlagFilename("in")
+	cmd.MarkFlagFilename("out")
+
+	return cmd
+}
+
+func (o *convert3Options) validate(args []string) error {
+	if len(o.Source) == 0 && o.SourceReader == nil {
+		return errors.New("an input file, or reader is required")
+	}
+
+	if len(o.Destination) == 0 && o.DestWriter == nil {
+		return errors.New("an output file or writer is required")
+	}
+	return nil
+}
+
+func (o *convert3Options) run() error {
+	var data []byte
+	switch {
+	case len(o.Source) > 0:
+		d, err := ioutil.ReadFile(o.Source)
+		if err != nil {
+			return err
+		}
+		data = d
+	case o.SourceReader != nil:
+		d, err := ioutil.ReadAll(o.SourceReader)
+		if err != nil {
+			return err
+		}
+		data = d
+	}
+
+	cfg, rpt, err := ignv2.Parse(data)
+	fmt.Fprintf(os.Stderr, "%s", rpt.String())
+	if err != nil || rpt.IsFatal() {
+		return errors.Errorf("Error parsing spec v2 config: %v\n%v", err, rpt)
+	}
```

Comment:
If the report contains warnings, we may as well print them, I think?

Reply:
Wait actually aren't we doing this with the `fmt.Fprintf(os.Stderr, "%s", rpt.String())` above?

Just tested this, yes we are.

Reply:
Whoops, no idea how I missed that.